x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -1011,22 +1011,34 @@ private AbstractRunnable fillGapRunnable(
             @Nullable SourceInputStreamFactory streamFactory,
             ActionListener<Void> listener
         ) {
-            return ActionRunnable.run(listener.delegateResponse((l, e) -> failGapAndListener(gap, l, e)), () -> {
-                var ioRef = io;
-                assert regionOwners.get(ioRef) == CacheFileRegion.this;
-                assert CacheFileRegion.this.hasReferences() : CacheFileRegion.this;
-                int start = Math.toIntExact(gap.start());
-                writer.fillCacheRange(
-                    ioRef,
-                    start,
-                    streamFactory,
-                    start,
-                    Math.toIntExact(gap.end() - start),
-                    progress -> gap.onProgress(start + progress)
-                );
-                writeCount.increment();
-                gap.onCompletion();
-            });
+            return new AbstractRunnable() {
+
+                @Override
+                protected void doRun() throws Exception {
+                    var ioRef = io;
+                    assert regionOwners.get(ioRef) == CacheFileRegion.this;
+                    assert CacheFileRegion.this.hasReferences() : CacheFileRegion.this;
+                    int start = Math.toIntExact(gap.start());
+                    writer.fillCacheRange(
+                        ioRef,
+                        start,
+                        streamFactory,
+                        start,
+                        Math.toIntExact(gap.end() - start),
+                        progress -> gap.onProgress(start + progress),
+                        listener.<Void>map(unused -> {
+                            writeCount.increment();
+                            gap.onCompletion();
+                            return null;
+                        }).delegateResponse((l, e) -> failGapAndListener(gap, l, e))
+                    );
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    failGapAndListener(gap, listener, e);
```

Comment:
I think that the contract of `fillCacheRange` changed now and it won't throw never? could you clarify?

Reply:
That's right, we can just delegate the failure to the listener.